firebase-database/src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java

```diff
@@ -344,6 +388,64 @@ public void listen(
     doIdleCheck();
   }
 
+  @Override
+  public Task<Object> get(List<String> path, Map<String, Object> queryParams) {
+    QuerySpec query = new QuerySpec(path, queryParams);
+    TaskCompletionSource<Object> source = new TaskCompletionSource<>();
+
+    long readId = this.readCounter++;
+
+    Map<String, Object> request = new HashMap<String, Object>();
+    request.put(REQUEST_PATH, ConnectionUtils.pathToString(query.path));
+    request.put(REQUEST_QUERIES, query.queryParams);
+
+    outstandingGets.put(
+        readId,
+        new OutstandingGet(
+            REQUEST_ACTION_GET,
+            request,
+            new ConnectionRequestCallback() {
+              @Override
+              public void onResponse(Map<String, Object> response) {
+                String status = (String) response.get(REQUEST_STATUS);
+                if (status.equals("ok")) {
+                  Object body = response.get(SERVER_DATA_UPDATE_BODY);
+                  delegate.onDataUpdate(query.path, body, /*isMerge=*/ false, /*tagNumber=*/ null);
+                  source.setResult(body);
+                } else {
+                  source.setException(
+                      new Exception((String) response.get(SERVER_DATA_UPDATE_BODY)));
+                }
+              }
+            }));
+
+    if (!connected()) {
+      executorService.schedule(
+          new Runnable() {
+            @Override
+            public void run() {
+              OutstandingGet get = outstandingGets.get(readId);
+              if (get == null || !get.markSent()) {
```

Comment:
It took me a while to wade through all the callsites, but it seems like `executorService` is single-threaded and the is responsible for all access to `outstandingGets`. This means that `markSent` doesn't need to use an AtomicBoolean.